euphoria-spark/src/main/java/cz/seznam/euphoria/spark/SparkFlowTranslator.java

```diff
@@ -51,114 +52,101 @@
-            SparkOperatorTranslator<O> translator, UnaryPredicate<O> accept) {
-      this.translator = Objects.requireNonNull(translator);
-      this.accept = accept;
-    }
-
-    static <O extends Operator<?, ?>> void set(
-            Map<Class, Translation> idx,
-            Class<O> type, SparkOperatorTranslator<O> translator)
-    {
-      set(idx, type, translator, null);
-    }
-
-    static <O extends Operator<?, ?>> void set(
-            Map<Class, Translation> idx,
-            Class<O> type, SparkOperatorTranslator<O> translator, UnaryPredicate<O> accept)
-    {
-      idx.put(type, new Translation<>(translator, accept));
-    }
-  }
+class SparkFlowTranslator {
 
   /* mapping of Euphoria operators to corresponding Flink transformations */
-  private final Map<Class, Translation> translations = new IdentityHashMap<>();
+  private final Map<Class, List<Translation>> translations = new IdentityHashMap<>();
 
   private final JavaSparkContext sparkEnv;
   private final Settings settings;
   private final SparkAccumulatorFactory accumulatorFactory;
 
-  public SparkFlowTranslator(JavaSparkContext sparkEnv,
-                             Settings flowSettings,
-                             SparkAccumulatorFactory accumulatorFactory) {
+  SparkFlowTranslator(JavaSparkContext sparkEnv,
+                      Settings flowSettings,
+                      SparkAccumulatorFactory accumulatorFactory) {
     this.sparkEnv = Objects.requireNonNull(sparkEnv);
     this.settings = Objects.requireNonNull(flowSettings);
     this.accumulatorFactory = Objects.requireNonNull(accumulatorFactory);
 
-    // basic operators
-    Translation.set(translations, FlowUnfolder.InputOperator.class, new InputTranslator());
-    Translation.set(translations, FlatMap.class, new FlatMapTranslator());
-    Translation.set(translations, ReduceStateByKey.class, new ReduceStateByKeyTranslator(flowSettings));
-    Translation.set(translations, Union.class, new UnionTranslator());
+    // ~ basic operators
+    Translation.add(translations, FlowUnfolder.InputOperator.class, new InputTranslator());
+    Translation.add(translations, FlatMap.class, new FlatMapTranslator());
+    Translation.add(translations, ReduceStateByKey.class, new ReduceStateByKeyTranslator(settings));
+    Translation.add(translations, Union.class, new UnionTranslator());
+
+    // ~ derived operators
+    Translation.add(translations, ReduceByKey.class, new ReduceByKeyTranslator(),
+        ReduceByKeyTranslator::wantTranslate);
 
-    // derived operators
-    Translation.set(translations, ReduceByKey.class, new ReduceByKeyTranslator(),
-            ReduceByKeyTranslator::wantTranslate);
+    // ~ batch broadcast join for a very small left side
+    Translation.add(translations, Join.class, new BroadcastJoinTranslator(), o ->
+        o.getHints().contains(JoinHints.broadcastHashJoin())
+            && (o.getWindowing() == null || o.getWindowing() instanceof GlobalWindowing));
```

Comment:
This look odd. Why do we constraint usability of this hint to to global or attached windowing? There should be no reason for that.